### PR TITLE
Toolmod migration fixes

### PR DESCRIPTION
**Pull request by mutability**
See #18797, #18799, #18805 for background.
This PR is a bit more conservative about fixing stuff than #18805

### CODE REVIEW

File: src/savegame_json.cpp [hunk 1]
```diff
@@ -1529,21 +1529,62 @@ void item::io( Archive& archive )
...
 
-        } else if( has_flag( "DOUBLE_REACTOR" ) ) {
-            item_tags.erase( "DOUBLE_REACTOR" );
-            emplace_back( "double_plutonium_core" );
+        }
+    }
+
+    // Fix fallout from #18797, which exponentially duplicates migrated toolmods
+    if( is_toolmod() ) {
+        // duplication would add an extra toolmod inside each toolmod on load;
```
Comment (mugling) [inline src/savegame_json.cpp]: This is probably overly paranoid. You could probably just clear the `contents` of any `item` where `is_toolmod()` returns true. Toolmods aren't supposed to contain anything

File: src/savegame_json.cpp [hunk 2]
```diff
@@ -1529,21 +1529,62 @@ void item::io( Archive& archive )
 
     // Migrate legacy toolmod flags
     if( !is_toolmod() ) {
-        if( has_flag( "ATOMIC_AMMO" ) ) {
```
Comment (mugling) [inline src/savegame_json.cpp]: This section here is superior to my approach as we don't end up with a `TOOLMOD_MIGRATION` flag applied to every tool

File: src/savegame_json.cpp [hunk 3]
```diff
@@ -1529,21 +1529,62 @@ void item::io( Archive& archive )
...
+        if( typeId() == "battery_atomic" || typeId() == "battery_compartment" ||
+            typeId() == "battery_ups" || typeId() == "double_plutonium_core" ) {
+            // Be conservative and only delete nested mods of the same type
+            contents.remove_if( [this]( const item &cont ) {
+                    return cont.typeId() == typeId();
+                        } );
+        }
+    }
+
+    if( is_tool() ) {
```
Comment (mugling) [inline src/savegame_json.cpp]: Arguably you could remove everything except the first toolmod which would be simpler code. That said you could just refactor this entire migration code into a function and call it at which point the cautious approach doesn't harm readbility

File: src/savegame_json.cpp [hunk 4]
```diff
@@ -1528,22 +1530,71 @@ void item::io( Archive& archive )
...
+            n_ups = 0,
+            n_plutonium = 0;
+
+        // not safe to use remove_if with a stateful predicate
+        for( auto it = i.contents.begin(); it != i.contents.end(); ) {
+            if( ( it->typeId() == "battery_atomic" && ++n_atomic > 1 ) ||
+                ( it->typeId() == "battery_compartment" && ++n_compartment > 1 ) ||
+                ( it->typeId() == "battery_ups" && ++n_ups > 1 ) ||
+                ( it->typeId() == "double_plutonium_core" && ++n_plutonium > 1 ) ) {
+                i.contents.erase( it++ );
```
Comment (mugling) [inline src/savegame_json.cpp]: Do you need to assign the return value of `erase` here?

Comment (mutability) [inline src/savegame_json.cpp]: Either postfix increment or assign the result, one or the other, which is preferred?

File: src/savegame_json.cpp [hunk 5]
```diff
@@ -1528,22 +1530,71 @@ void item::io( Archive& archive )
...
+            n_ups = 0,
+            n_plutonium = 0;
+
+        // not safe to use remove_if with a stateful predicate
+        for( auto i = it.contents.begin(); i != it.contents.end(); ) {
+            if( ( i->typeId() == "battery_atomic" && ++n_atomic > 1 ) ||
+                ( i->typeId() == "battery_compartment" && ++n_compartment > 1 ) ||
+                ( i->typeId() == "battery_ups" && ++n_ups > 1 ) ||
+                ( i->typeId() == "double_plutonium_core" && ++n_plutonium > 1 ) ) {
+                it.contents.erase( i++ );
```
Comment (mugling) [inline src/savegame_json.cpp]: Should this not be
`i = it.contents.erase( i )`?

Comment (mutability) [inline src/savegame_json.cpp]: Either are valid. Assigning the result is the C++11 idiom, postincrementing the iterator is the older idiom (pre-C++11, not all container types returned an iterator from erase). My C++ dates back a decade or so, so my muscle memory goes for the postincrement. Which style is preferred?

Comment (mugling) [inline src/savegame_json.cpp]: Unless you are certain they are identical I'd prefer the form we know works in all cases (assignment of returned iterator)

Comment (mutability) [inline src/savegame_json.cpp]: I am certain they are identical, but if the existing convention in the code is to assign the iterator then I'll make that change.

(the postincrement does: take a temporary copy of the iterator, increment the iterator, pass the temporary copy to erase; erase invalidates the iterator it was passed, but we have already moved 'it' to the next item before that happens)

Comment (mugling) [inline src/savegame_json.cpp]: Always good to learn something new. If they are identical I actually prefer your variant from a syntax POV

**Pull request merged**